File: src/services/sales.ts
import type { InvoiceItemRow, InvoiceRow, PaymentStatus, InvoiceState } from "@/shared/domain/types";
import { API_ROUTES } from "@/shared/constants/api";
import { MAX_PAGE_SIZE } from "@/shared/constants/config";
import { apiRequest, queryString, type PaginatedResponse } from "@/services/api";

export interface InvoiceItemResponse {
  id: string;
  productId: string;
  productUnitId?: string | null;
  productName: string;
  sku: string;
  unitName: string;
  displayedQuantity: number;
  conversionToBase: number;
  baseQuantity: number;
  unitPrice: number;
  minimumPriceSnapshot: number;
  priceMode: string;
  soldBelowMinimum: boolean;
  authorizedBy?: string | null;
  discount: number;
  tax: number;
  lineTotal: number;
  fifoCost: number;
  grossProfit: number;
}

export interface InvoiceResponse {
  id: string;
  branchId: string;
  customerId?: string | null;
  invoiceNumber: string;
  status: string;
  paymentStatus: string;
  subtotal: number;
  discount: number;
  tax: number;
  total: number;
  paidAmount: number;
  creditAmount: number;
  changeAmount: number;
  notes?: string | null;
  cashierName: string;
  clientRequestId: string;
  items: InvoiceItemResponse[];
  payments: unknown[];
  completedAt?: string | null;
  voidReason?: string | null;
  voidedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface HoldResponse {
  id: string;
  label: string;
  customerId?: string | null;
  payload: unknown;
  expiresAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CompleteSalePayload {
  clientRequestId: string;
  customerId?: string | null;
  items: {
    productId: string;
    productUnitId: string;
    quantity: number;
    unitPrice: number;
    priceMode?: string;
    soldBelowMinimum?: boolean;
    authorizedBy?: string | null;
    discount?: number;
    tax?: number;
  }[];
  payments: {
    amount: number;
    amountTendered?: number | null;
    paymentMethod?: string;
    referenceNumber?: string | null;
    notes?: string | null;
    clientRequestId?: string | null;
  }[];
  discount?: number;
  tax?: number;
  notes?: string | null;
}

export function mapInvoice(row: InvoiceResponse): InvoiceRow {
  const items: InvoiceItemRow[] = (row.items ?? []).map((item) => ({
    id: item.id,
    productId: item.productId,
    unitName: item.unitName,
    quantity: item.displayedQuantity,
    baseQuantity: item.baseQuantity,
    unitPrice: item.unitPrice,
    discount: item.discount,
    total: item.lineTotal,
  }));
  return {
    id: row.id,
    invoiceNumber: row.invoiceNumber,
    branchId: row.branchId,
    customerId: row.customerId ?? null,
    subtotal: row.subtotal,
    discount: row.discount,
    total: row.total,
    paidAmount: row.paidAmount,
    creditAmount: row.creditAmount,
    paymentStatus: (row.paymentStatus as PaymentStatus) || "UNPAID",
    status: (row.status as InvoiceState) || "COMPLETED",
    notes: row.notes ?? "",
    createdBy: row.cashierName,
    createdAt: row.createdAt,
    items,
  };
}

export function listInvoices(
  params: {
    page?: number;
    perPage?: number;
    branchId?: string;
    customerId?: string;
    status?: string;
    paymentStatus?: string;
  } = {},
  signal?: AbortSignal,
) {
  return apiRequest<PaginatedResponse<InvoiceResponse>>(
    `${API_ROUTES.invoices}${queryString(params)}`,
    { signal },
  );
}

export async function listAllInvoices(signal?: AbortSignal): Promise<InvoiceRow[]> {
  const rows: InvoiceRow[] = [];
  let page = 1;
  for (;;) {
    const response = await listInvoices({ page, perPage: MAX_PAGE_SIZE }, signal);
    rows.push(...response.data.map(mapInvoice));
    if (!response.meta.hasNextPage) return rows;
    page += 1;
  }
}

export function getInvoice(id: string, signal?: AbortSignal) {
  return apiRequest<InvoiceResponse>(API_ROUTES.invoiceById(id), { signal }).then(mapInvoice);
}

export function completeSale(payload: CompleteSalePayload) {
  return apiRequest<InvoiceResponse>(API_ROUTES.salesComplete, {
    method: "POST",
    body: JSON.stringify(payload),
  }).then(mapInvoice);
}

export function voidInvoice(id: string, reason?: string) {
  return apiRequest<InvoiceResponse>(API_ROUTES.invoiceVoid(id), {
    method: "POST",
    body: JSON.stringify({ reason: reason ?? "" }),
  }).then(mapInvoice);
}

export function listHolds(signal?: AbortSignal) {
  return apiRequest<HoldResponse[] | PaginatedResponse<HoldResponse>>(
    API_ROUTES.salesHolds,
    { signal },
  ).then((payload) => (Array.isArray(payload) ? payload : payload.data ?? []));
}

export function createHold(payload: {
  label: string;
  customerId?: string | null;
  payload: unknown;
  expiresAt?: string | null;
}) {
  return apiRequest<HoldResponse>(API_ROUTES.salesHolds, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function deleteHold(id: string) {
  return apiRequest<{ id: string; deleted: boolean }>(API_ROUTES.salesHoldById(id), {
    method: "DELETE",
  });
}

File: src-tauri/src/backend/dto/sales.rs
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use validator::{Validate, ValidationError};

use super::PageQuery;

#[derive(Clone, Debug, Deserialize, Validate)]
#[validate(schema(function = "validate_complete_sale"))]
#[serde(rename_all = "camelCase")]
pub struct CompleteSaleRequest {
    #[validate(length(min = 1, max = 120))]
    pub client_request_id: String,
    pub customer_id: Option<String>,
    #[validate(nested)]
    pub items: Vec<SaleItemInput>,
    #[validate(nested)]
    pub payments: Vec<SalePaymentInput>,
    #[validate(custom(function = "non_negative_decimal"))]
    #[serde(default)]
    pub discount: Decimal,
    #[validate(custom(function = "non_negative_decimal"))]
    #[serde(default)]
    pub tax: Decimal,
    pub notes: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct SaleItemInput {
    pub product_id: String,
    pub product_unit_id: String,
    #[validate(custom(function = "positive_decimal"))]
    pub quantity: Decimal,
    #[validate(custom(function = "non_negative_decimal"))]
    pub unit_price: Decimal,
    #[serde(default = "default_price_mode")]
    pub price_mode: String,
    #[serde(default)]
    pub sold_below_minimum: bool,
    pub authorized_by: Option<String>,
    #[validate(custom(function = "non_negative_decimal"))]
    #[serde(default)]
    pub discount: Decimal,
    #[validate(custom(function = "non_negative_decimal"))]
    #[serde(default)]
    pub tax: Decimal,
}

#[derive(Clone, Debug, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct SalePaymentInput {
    #[validate(custom(function = "positive_decimal"))]
    pub amount: Decimal,
    pub amount_tendered: Option<Decimal>,
    #[serde(default = "default_payment_method")]
    pub payment_method: String,
    pub reference_number: Option<String>,
    pub notes: Option<String>,
    pub client_request_id: Option<String>,
}

#[derive(Clone, Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct InvoiceListQuery {
    #[serde(flatten)]
    pub page: PageQuery,
    #[serde(alias = "branch_id")]
    pub branch_id: Option<String>,
    #[serde(alias = "customer_id")]
    pub customer_id: Option<String>,
    pub status: Option<String>,
    #[serde(alias = "payment_status")]
    pub payment_status: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InvoiceItemResponse {
    pub id: String,
    pub product_id: String,
    pub product_unit_id: Option<String>,
    pub product_name: String,
    pub sku: String,
    pub unit_name: String,
    pub displayed_quantity: f64,
    pub conversion_to_base: f64,
    pub base_quantity: f64,
    pub unit_price: f64,
    pub minimum_price_snapshot: f64,
    pub price_mode: String,
    pub sold_below_minimum: bool,
    pub authorized_by: Option<String>,
    pub discount: f64,
    pub tax: f64,
    pub line_total: f64,
    pub fifo_cost: f64,
    pub gross_profit: f64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InvoicePaymentResponse {
    pub id: String,
    pub amount: f64,
    pub amount_tendered: Option<f64>,
    pub change_amount: f64,
    pub payment_method: String,
    pub reference_number: Option<String>,
    pub status: String,
    pub paid_at: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InvoiceResponse {
    pub id: String,
    pub branch_id: String,
    pub customer_id: Option<String>,
    pub invoice_number: String,
    pub status: String,
    pub payment_status: String,
    pub subtotal: f64,
    pub discount: f64,
    pub tax: f64,
    pub total: f64,
    pub paid_amount: f64,
    pub credit_amount: f64,
    pub change_amount: f64,
    pub notes: Option<String>,
    pub cashier_name: String,
    pub client_request_id: String,
    pub items: Vec<InvoiceItemResponse>,
    pub payments: Vec<InvoicePaymentResponse>,
    pub completed_at: Option<String>,
    pub void_reason: Option<String>,
    pub voided_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone, Debug, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct HoldRequest {
    #[validate(length(min = 1, max = 120))]
    pub label: String,
    pub customer_id: Option<String>,
    pub payload: serde_json::Value,
    pub expires_at: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HoldResponse {
    pub id: String,
    pub branch_id: String,
    pub customer_id: Option<String>,
    pub label: String,
    pub payload: serde_json::Value,
    pub held_by: String,
    pub held_at: String,
    pub expires_at: Option<String>,
    pub updated_at: String,
}

#[derive(Clone, Debug, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct VoidRequest {
    #[validate(length(min = 1, max = 500))]
    pub reason: String,
}

fn default_price_mode() -> String {
    "RETAIL".into()
}

fn default_payment_method() -> String {
    "CASH".into()
}

fn positive_decimal(value: &Decimal) -> Result<(), ValidationError> {
    (*value > Decimal::ZERO)
        .then_some(())
        .ok_or_else(|| ValidationError::new("must_be_positive"))
}

fn non_negative_decimal(value: &Decimal) -> Result<(), ValidationError> {
    (*value >= Decimal::ZERO)
        .then_some(())
        .ok_or_else(|| ValidationError::new("must_be_non_negative"))
}

fn validate_complete_sale(value: &CompleteSaleRequest) -> Result<(), ValidationError> {
    if value.items.is_empty() {
        return Err(ValidationError::new("empty_sale"));
    }
    for item in &value.items {
        if item.sold_below_minimum
            && item
                .authorized_by
                .as_deref()
                .map(str::trim)
                .filter(|v| !v.is_empty())
                .is_none()
        {
            return Err(ValidationError::new("below_minimum_unauthorized"));
        }
        let mode = item.price_mode.trim().to_uppercase();
        if mode != "RETAIL" && mode != "WHOLESALE" {
            return Err(ValidationError::new("invalid_price_mode"));
        }
    }
    Ok(())
}
